File: apps/gyst/tests/pty/editor.ts
// Linux real-PTY regression; requires the pinned Bun runtime and Vim. Evidence stays outside the repo.
import assert from "node:assert/strict";
import { existsSync, mkdirSync, mkdtempSync, readFileSync, rmSync, writeFileSync } from "node:fs";
import { constants, tmpdir } from "node:os";
import { join, resolve } from "node:path";
import { parseArgs } from "node:util";

function screen(data: Buffer) {
  let cells = Array.from({ length: 40 }, () => Array<string>(240).fill(" "));
  let row = 0;
  let col = 0;
  // OSC terminates at BEL or ST, never at a BEL in a later frame.
  const tokens = data.toString().matchAll(
    // eslint-disable-next-line no-control-regex -- Interpret actual terminal control sequences.
    /\x1b\[[0-?]*[ -/]*[@-~]|\x1b(?:\][^\x07\x1b]*(?:\x07|\x1b\\)|[P_].*?\x1b\\)|[^\x1b]/gs,
  );
  for (const [token] of tokens) {
    if (token.startsWith("\x1b[")) {
      const cmd = token.at(-1)!;
      const values = token
        .slice(2, -1)
        .split(";")
        .map((n) => (/^\d+$/.test(n) ? Number(n) : 0));
      const first = values[0] || 1;
      if (cmd === "H" || cmd === "f") {
        row = first - 1;
        col = (values[1] || 1) - 1;
      } else if (cmd === "A") row -= first;
      else if (cmd === "B") row += first;
      else if (cmd === "C") col += first;
      else if (cmd === "D") col -= first;
      else if (cmd === "G") col = first - 1;
      else if (cmd === "d") row = first - 1;
      else if (cmd === "J" && [2, 3].includes(values[0]!))
        cells = Array.from({ length: 40 }, () => Array<string>(240).fill(" "));
      else if (cmd === "K" && cells[row]) {
        const start = [1, 2].includes(values[0]!) ? 0 : Math.max(0, col);
        const end = [0, 2].includes(values[0]!) ? 240 : Math.min(240, col + 1);
        cells[row]!.fill(" ", start, end);
      }
    } else if (token.startsWith("\x1b")) continue;
    else if (token === "\r") col = 0;
    else if (token === "\n") row++;
    else if (token >= " ") {
      if (cells[row] && col >= 0 && col < 240) cells[row]![col] = token;
      col++;
    }
  }
  return cells.map((line) => line.join("")).join("\n");
}

type Event = { event: string; [key: string]: unknown };
type Child = {
  pid: number;
  argv?: string[];
  cwd?: string;
  tty?: boolean;
  canonical?: boolean;
  pgrp?: number;
  foreground?: number;
  descendant?: number;
};
const termination: Record<string, NodeJS.Signals> = {
  term: "SIGTERM",
  hup: "SIGHUP",
  "quit-signal": "SIGQUIT",
  "ignore-term": "SIGTERM",
  "descendant-term": "SIGTERM",
  abrt: "SIGABRT",
  pipe: "SIGPIPE",
  bus: "SIGBUS",
  "vim-term": "SIGTERM",
};
// Linux termios c_lflag bits; this harness deliberately claims Linux runtime coverage only.
const ICANON = 0x2;
const ECHO = 0x8;

async function runCase(name: string, evidence: string, vim: string) {
  const root = mkdtempSync(join(tmpdir(), "gyst editor pty "));
  const target = join(root, "target with spaces.txt");
  const log = join(root, "events.jsonl");
  const childLog = join(root, "child.json");
  const editor = join(root, "editor with spaces");
  writeFileSync(target, "before\n");
  writeFileSync(
    editor,
    `#!${process.execPath}\nawait import(${JSON.stringify(join(import.meta.dir, "editor-fixture.ts"))});\n`,
    { mode: 0o755 },
  );
  const env: NodeJS.ProcessEnv = {
    ...process.env,
    EDITOR: editor,
    TERM: "xterm-256color",
    GYST_PTY_LOG: log,
    GYST_PTY_CHILD: childLog,
    GYST_PTY_MODE: name,
    GYST_PTY_VIM: vim,
    GYST_PTY_SOURCE: name === "stdin" ? "stdin" : "git",
    HOME: root,
    XDG_CONFIG_HOME: join(root, "config"),
  };
  if (name === "missing") env.EDITOR = "gyst-no-such-editor";
  if (name === "unset") delete env.EDITOR;
  const captured: Buffer[] = [];
  const terminal = new Bun.Terminal({
    cols: 120,
    rows: 30,
    data: (_, data) => captured.push(Buffer.from(data)),
  });
  const baseline = terminal.localFlags;
  // Bun allocates the PTY; setsid assigns its controlling session for terminal-generated signals.
  const driver = Bun.spawn(
    [
      "setsid",
      "--ctty",
      process.execPath,
      "--preload",
      resolve(import.meta.dir, "../../node_modules/@opentui/solid/scripts/preload.js"),
      join(import.meta.dir, "editor-driver.tsx"),
    ],
    { cwd: root, env, terminal },
  );
  let child: Child | undefined;
  const bytes = () => Buffer.concat(captured);
  const events = (): Event[] =>
    existsSync(log)
      ? readFileSync(log, "utf8")
          .split("\n")
          .filter((line) => line.endsWith("}"))
          .map((line) => JSON.parse(line))
      : [];
  const event = (kind: string) => events().filter((entry) => entry.event === kind);
  const wait = async (predicate: () => boolean, description: string) => {
    const deadline = performance.now() + 8_000;
    while (!predicate()) {
      assert(
        performance.now() < deadline,
        JSON.stringify({
          name,
          description,
          exit: driver.exitCode,
          events: events(),
          output: bytes().toString().slice(-1000),
        }),
      );
      await Bun.sleep(25);
    }
  };
  const send = (value: string) => terminal.write(value);
  try {
    await wait(() => event("mounted").length > 0, "mount");
    await wait(() => screen(bytes()).includes("[diff]"), "initial frame");
    send("o");
    if (!["missing", "unset", "suspend"].includes(name)) {
      await wait(
        () => existsSync(childLog) && readFileSync(childLog, "utf8").trimEnd().endsWith("}"),
        "editor start",
      );
      child = JSON.parse(readFileSync(childLog, "utf8")) as Child;
      if (!name.startsWith("vim")) {
        assert.deepEqual(child.argv, [target]);
        assert.equal(child.cwd, root);
        assert(child.tty);
        assert(child.canonical, "editor receives canonical input");
        assert.equal(child.pgrp, child.foreground, "child retains terminal foreground group");
      }
      if (termination[name] && name !== "vim-term") driver.kill(termination[name]);
      else if (name === "destroy") driver.kill("SIGUSR1");
      else if (["ctrl-c", "ignore-int", "descendant-int"].includes(name)) send("\x03");
      else if (name.startsWith("vim")) {
        for (const width of [80, 200, 120]) {
          terminal.resize(width, 30);
          await Bun.sleep(80);
        }
        if (name === "vim-save") send("GoREAL_EDITOR_SAVE\x1b:wq\r");
        else if (name === "vim-term") driver.kill("SIGTERM");
        else {
          send("\x03");
          await Bun.sleep(100);
          assert.equal(
            event("returned").length,
            0,
            "Vim must own Ctrl+C; never type into the resumed TUI",
          );
          send(":q!\r");
        }
      } else if (!["nonzero", "noop"].includes(name)) {
        for (const width of [80, 200, 120]) {
          terminal.resize(width, 30);
          await Bun.sleep(50);
        }
        send("x\n");
        if (name === "backlog") send("q\n");
      }
    }
    const shuttingDown = Boolean(termination[name]) || ["destroy", "resume"].includes(name);
    await wait(() => event("returned").length > 0, "handoff return");
    const returned = event("returned")[0]!;
    assert.equal(returned.refreshes, 0);
    assert.equal(returned.reads, event("edit")[0]!.reads, "no polls during handoff");
    if (shuttingDown) {
      await wait(() => driver.exitCode !== null, "terminated app");
      assert.equal(event("resumed").length, 0, "shutdown never resumes");
      assert.equal(
        driver.exitCode,
        termination[name] ? 128 + constants.signals[termination[name]] : 0,
      );
    } else {
      assert.deepEqual(returned.listeners, returned.before, "handoff listeners removed");
      assert(returned.raw && !returned.destroyed);
      assert.equal(event("resumed").length, ["missing", "unset"].includes(name) ? 0 : 1);
      const guidance = name === "stdin" ? "stdin snapshot unchanged" : "snapshot unchanged";
      await wait(() => screen(bytes()).includes(guidance), "snapshot guidance");
      await Bun.sleep(100);
      assert(
        driver.exitCode === null && event("quit").length === 0,
        "no replay of editor keystrokes",
      );
      send("q");
      await wait(() => driver.exitCode !== null, "normal quit");
      assert.equal(driver.exitCode, 0);
    }
    await driver.exited;
    await wait(() => event("finished").length > 0, "outer finalizer");
    const finished = event("finished")[0]!;
    assert.equal(finished.refreshes, 0);
    assert.deepEqual(finished.listeners, finished.initialListeners);
    assert(terminal.localFlags & ICANON, "canonical terminal restored");
    assert.equal(terminal.localFlags & ECHO, baseline & ECHO, "echo restored");
    await Bun.sleep(50);
    const output = bytes().toString();
    assert.equal(
      output.split("\x1b[?1049h").length,
      output.split("\x1b[?1049l").length,
      "alternate screen balanced",
    );
    assert(!event("resuming").some((entry) => entry.destroyed), "no resume after destroy");
    for (const pid of [child?.pid, child?.descendant]) {
      if (pid !== undefined)
        assert.throws(() => process.kill(pid, 0), { code: "ESRCH" }, "child reaped");
    }
    if (name === "save") assert.equal(readFileSync(target, "utf8"), "saved by fake editor\n");
    if (name === "vim-save") assert(readFileSync(target, "utf8").includes("REAL_EDITOR_SAVE"));
    console.log(
      `PASS ${name}: exit=${driver.exitCode}, resumes=${event("resumed").length}, terminal restored, child reaped`,
    );
  } finally {
    if (driver.exitCode === null) {
      driver.kill("SIGTERM");
      await Promise.race([driver.exited, Bun.sleep(2_000)]);
      if (driver.exitCode === null) {
        driver.kill("SIGKILL");
        await driver.exited;
      }
    }
    for (const pid of [child?.pid, child?.descendant]) {
      if (pid === undefined) continue;
      try {
        process.kill(pid, "SIGKILL");
      } catch (error) {
        assert.equal((error as NodeJS.ErrnoException).code, "ESRCH");
      }
    }
    try {
      writeFileSync(join(evidence, `${name}.terminal.bin`), bytes());
      writeFileSync(join(evidence, `${name}.events.json`), JSON.stringify(events(), null, 2));
    } finally {
      terminal.close();
      rmSync(root, { recursive: true, force: true });
    }
  }
}

assert.equal(
  process.platform,
  "linux",
  "This harness verifies Linux; request other runtime access separately.",
);
const { values } = parseArgs({ options: { evidence: { type: "string" } }, strict: true });
assert(values.evidence, "--evidence /path/outside/repo is required");
const vim = Bun.which("vim");
assert(
  vim && Bun.which("stty") && Bun.which("setsid"),
  "Vim, stty and setsid are required; do not silently skip editor verification.",
);
mkdirSync(values.evidence, { recursive: true });
for (const name of [
  "noop",
  "save",
  "stdin",
  "nonzero",
  "missing",
  "unset",
  "ctrl-c",
  "ignore-int",
  "backlog",
  "term",
  "hup",
  "quit-signal",
  "ignore-term",
  "abrt",
  "pipe",
  "bus",
  "descendant-int",
  "descendant-term",
  "destroy",
  "suspend",
  "resume",
  "vim-noop",
  "vim-save",
  "vim-term",
])
  await runCase(name, resolve(values.evidence), vim);

File: apps/gyst/tests/pty/editor.py
#!/usr/bin/env python3
"""Linux real-PTY regression: python3 tests/pty/editor.py --evidence /outside/repo/path.
Requires Bun and Vim on PATH. Uses only stdlib, isolated sessions, continuous drain,
and bounded cleanup. It does not claim macOS/Windows runtime coverage.
"""
import argparse
import fcntl
import json
import os
from pathlib import Path
import re
import select
import shutil
import signal
import struct
import subprocess
import sys
import tempfile
import termios
import threading
import time


def screen(data):
    cells = [[" "] * 240 for _ in range(40)]
    row = col = 0
    for match in re.finditer(r"\x1b\[[0-?]*[ -/]*[@-~]|\x1b(?:\][^\x07]*(?:\x07|\x1b\\)|[P_].*?\x1b\\)|[^\x1b]", data.decode(errors="ignore"), re.S):
        token = match.group()
        if token.startswith("\x1b["):
            cmd = token[-1]
            values = [int(n) if n.isdigit() else 0 for n in token[2:-1].split(";")]
            first = values[0] or 1
            if cmd in "Hf": row, col = first - 1, (values[1] or 1) - 1 if len(values) > 1 else 0
            elif cmd == "A": row -= first
            elif cmd == "B": row += first
            elif cmd == "C": col += first
            elif cmd == "D": col -= first
            elif cmd == "G": col = first - 1
            elif cmd == "d": row = first - 1
            elif cmd == "J" and values[0] in [2, 3]: cells = [[" "] * 240 for _ in range(40)]
            elif cmd == "K" and 0 <= row < 40:
                for c in range(0 if values[0] in [1, 2] else max(0, col), 240 if values[0] in [0, 2] else min(240, col + 1)): cells[row][c] = " "
        elif token.startswith("\x1b"): pass
        elif token == "\r": col = 0
        elif token == "\n": row += 1
        elif token >= " ":
            if 0 <= row < 40 and 0 <= col < 240: cells[row][col] = token
            col += 1
    return "\n".join("".join(line) for line in cells)


def run_case(name, evidence, bun, vim):
    with tempfile.TemporaryDirectory(prefix="gyst editor pty ") as directory:
        root = Path(directory)
        target = root / "target with spaces.txt"
        target.write_text("before\n")
        log = root / "events.jsonl"
        child_log = root / "child.json"
        editor = root / "editor with spaces"
        if name.startswith("vim"):
            # An explicit executable wrapper is the supported way to supply flags.
            editor.write_text(f"#!{sys.executable}\nimport os,json\nfrom pathlib import Path\nPath({str(child_log)!r}).write_text(json.dumps({{'pid':os.getpid()}}))\nos.execv({vim!r}, [{vim!r}, '-N', '-u', 'NONE', '-i', 'NONE', '-n', *os.sys.argv[1:]])\n")
        else:
            editor.write_text(f'''#!{sys.executable}
import json,os,signal,sys,time,termios,subprocess
from pathlib import Path
mode={name!r}
descendant = None
if mode.startswith('descendant'):
    descendant = subprocess.Popen([sys.executable, '-c', 'import time; time.sleep(30)'])
    def end(sig, _frame):
        descendant.send_signal(sig)
        descendant.wait(timeout=2)
        sys.exit(128 + sig)
    signal.signal(signal.SIGTERM, end)
    signal.signal(signal.SIGINT, end)
if mode in ['ignore-term', 'ignore-int']:
    signal.signal(signal.SIGTERM if mode == 'ignore-term' else signal.SIGINT, lambda *_: None)
Path({str(child_log)!r}).write_text(json.dumps({{'pid':os.getpid(),'argv':sys.argv[1:],'cwd':os.getcwd(),'tty':all(os.isatty(i) for i in [0,1,2]),'canonical':bool(termios.tcgetattr(0)[3] & termios.ICANON),'pgrp':os.getpgrp(),'foreground':os.tcgetpgrp(0),'descendant':descendant.pid if descendant else None}}))
if mode == 'nonzero': sys.exit(7)
if mode == 'noop': sys.exit(0)
if mode in ['ignore-term','ignore-int','descendant-int','descendant-term']:
    while True: time.sleep(.1)
try: sys.stdin.readline()
except KeyboardInterrupt: sys.exit(130)
if mode == 'backlog': time.sleep(.2)
if mode == 'save': Path(sys.argv[1]).write_text('saved by fake editor\\n')
''')
        editor.chmod(0o755)
        env = dict(os.environ, EDITOR=str(editor), TERM="xterm-256color", GYST_PTY_LOG=str(log), GYST_PTY_MODE=name,
                   GYST_PTY_SOURCE="stdin" if name == "stdin" else "git", HOME=str(root), XDG_CONFIG_HOME=str(root / "config"))
        if name == "missing": env["EDITOR"] = "gyst-no-such-editor"
        if name == "unset": env.pop("EDITOR", None)
        master, slave = os.openpty()
        baseline = termios.tcgetattr(slave)
        fcntl.ioctl(slave, termios.TIOCSWINSZ, struct.pack("HHHH", 30, 120, 0, 0))
        def terminal_session():
            os.setsid()
            fcntl.ioctl(0, termios.TIOCSCTTY, 0)
            os.tcsetpgrp(0, os.getpgrp())
        app = Path(__file__).resolve().parents[2]
        process = subprocess.Popen([bun, "--preload", str(app / "node_modules/@opentui/solid/scripts/preload.js"), str(Path(__file__).with_name("editor-driver.tsx").resolve())],
                                   cwd=root, env=env, stdin=slave, stdout=slave, stderr=slave, preexec_fn=terminal_session)
        captured = bytearray()
        stop = threading.Event()
        def drain():
            while not stop.is_set():
                if select.select([master], [], [], .025)[0]:
                    try:
                        data = os.read(master, 65536)
                        if not data: break
                        captured.extend(data)
                    except OSError: break
        reader = threading.Thread(target=drain, daemon=True)
        reader.start()
        child = None
        def events():
            if not log.exists(): return []
            return [json.loads(line) for line in log.read_text().splitlines() if line.endswith("}")]
        def event(kind): return [e for e in events() if e["event"] == kind]
        def wait(predicate, description, timeout=8):
            end = time.monotonic() + timeout
            while not predicate():
                assert time.monotonic() < end, (name, description, process.poll(), events(), bytes(captured)[-1000:])
                time.sleep(.025)
        def send(value): os.write(master, value)
        def resize(width):
            fcntl.ioctl(slave, termios.TIOCSWINSZ, struct.pack("HHHH", 30, width, 0, 0))
            os.kill(process.pid, signal.SIGWINCH)
        try:
            wait(lambda: event("mounted"), "mount")
            wait(lambda: "[diff]" in screen(bytes(captured)), "initial frame")
            send(b"o")
            if name not in ["missing", "unset", "suspend"]:
                wait(child_log.exists, "editor start")
                child = json.loads(child_log.read_text())
                if not name.startswith("vim"):
                    assert child["argv"] == [str(target)] and child["cwd"] == str(root)
                    assert child["tty"] and child["canonical"]
                    assert child["pgrp"] == child["foreground"], "child retains terminal foreground group"
                if name in ["term", "hup", "quit-signal", "ignore-term", "descendant-term", "abrt", "pipe", "bus"]:
                    os.kill(process.pid, {"term": signal.SIGTERM, "hup": signal.SIGHUP, "quit-signal": signal.SIGQUIT, "ignore-term": signal.SIGTERM, "descendant-term": signal.SIGTERM, "abrt": signal.SIGABRT, "pipe": signal.SIGPIPE, "bus": signal.SIGBUS}[name])
                elif name == "destroy": os.kill(process.pid, signal.SIGUSR1)
                elif name in ["ctrl-c", "ignore-int", "descendant-int"]: send(b"\x03")
                elif name.startswith("vim"):
                    wait(lambda: "before" in screen(bytes(captured)), "Vim file")
                    for width in [80, 200, 120]: resize(width); time.sleep(.08)
                    if name == "vim-save": send(b"GoREAL_EDITOR_SAVE\x1b:wq\r")
                    elif name == "vim-term": os.kill(process.pid, signal.SIGTERM)
                    else:
                        send(b"\x03")
                        time.sleep(.1)
                        send(b":q!\r")
                elif name not in ["nonzero", "noop"]:
                    for width in [80, 200, 120]: resize(width); time.sleep(.05)
                    send(b"x\n")
                    if name == "backlog": send(b"q\n")
            shutting_down = name in ["term", "hup", "quit-signal", "ignore-term", "destroy", "resume", "vim-term", "descendant-term", "abrt", "pipe", "bus"]
            wait(lambda: event("returned"), "handoff return")
            returned = event("returned")[0]
            assert returned["refreshes"] == 0
            assert returned["reads"] == event("edit")[0]["reads"], "no polls during handoff"
            if shutting_down:
                wait(lambda: process.poll() is not None, "terminated app")
                assert not event("resumed"), "shutdown never resumes"
                expected = {"term":143, "hup":129, "quit-signal":131, "ignore-term":143, "vim-term":143, "descendant-term":143, "abrt":134, "pipe":141, "bus":135}.get(name, 0)
                assert process.returncode == expected, (name, process.returncode)
            else:
                assert returned["listeners"] == returned["before"], "handoff listeners removed"
                assert returned["raw"] and not returned["destroyed"]
                assert len(event("resumed")) == (0 if name in ["missing", "unset"] else 1)
                guidance = "stdin snapshot unchanged" if name == "stdin" else "snapshot unchanged"
                wait(lambda: guidance in screen(bytes(captured)), "snapshot guidance")
                time.sleep(.1)
                assert process.poll() is None and not event("quit"), "no replay of editor keystrokes"
                send(b"q")
                wait(lambda: process.poll() is not None, "normal quit")
                assert process.returncode == 0
            wait(lambda: event("finished"), "outer finalizer")
            assert event("finished")[0]["refreshes"] == 0
            assert event("finished")[0]["listeners"] == event("finished")[0]["initialListeners"]
            assert termios.tcgetattr(slave)[3] & termios.ICANON, "canonical terminal restored"
            assert bool(termios.tcgetattr(slave)[3] & termios.ECHO) == bool(baseline[3] & termios.ECHO)
            assert bytes(captured).count(b"\x1b[?1049h") == bytes(captured).count(b"\x1b[?1049l"), "alternate screen balanced"
            assert not any(e.get("destroyed") for e in event("resuming")), "no resume after destroy"
            if child:
                for pid in [child["pid"], child.get("descendant")]:
                    if pid is None: continue
                    try: os.kill(pid, 0)
                    except ProcessLookupError: pass
                    else: raise AssertionError((name, "child not reaped", child))
            if name == "save": assert target.read_text() == "saved by fake editor\n"
            if name == "vim-save": assert "REAL_EDITOR_SAVE" in target.read_text()
            print(f"PASS {name}: exit={process.returncode}, resumes={len(event('resumed'))}, terminal restored, child reaped", flush=True)
        finally:
            if process.poll() is None:
                process.terminate()
                try: process.wait(timeout=2)
                except subprocess.TimeoutExpired:
                    process.kill()
                    process.wait(timeout=2)
            if child:
                for pid in [child["pid"], child.get("descendant")]:
                    if pid is None: continue
                    try: os.kill(pid, signal.SIGKILL)
                    except ProcessLookupError: pass
            time.sleep(.05)
            stop.set()
            reader.join(timeout=1)
            (evidence / f"{name}.terminal.bin").write_bytes(captured)
            (evidence / f"{name}.events.json").write_text(json.dumps(events(), indent=2))
            os.close(master)
            os.close(slave)


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--evidence", type=Path, required=True)
    args = parser.parse_args()
    assert sys.platform == "linux", "This harness verifies Linux; request other runtime access separately."
    bun, vim = shutil.which("bun"), shutil.which("vim")
    assert bun and vim, "Bun and real Vim are required; do not silently skip editor verification."
    args.evidence.mkdir(parents=True, exist_ok=True)
    for case in ["noop", "save", "stdin", "nonzero", "missing", "unset", "ctrl-c", "ignore-int", "backlog", "term", "hup", "quit-signal", "ignore-term", "abrt", "pipe", "bus", "descendant-int", "descendant-term", "destroy", "suspend", "resume", "vim-noop", "vim-save", "vim-term"]:
        run_case(case, args.evidence, bun, vim)
